src/events/upgrade.py

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Event handler for handling OpensearchDashboards in-place upgrades."""
+import logging
+import time
+from typing import TYPE_CHECKING
+
+from charms.data_platform_libs.v0.upgrade import (
+    BaseModel,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+)
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    UpgradeGrantedEvent,
+)
+from typing_extensions import override
+
+if TYPE_CHECKING:
+    from charm import (
+        OpensearchDashboardsCharm,  # pyright: ignore [reportGeneralTypeIssues, reportAttributeAccessIssue]
+    )
+
+logger = logging.getLogger(__name__)
+
+
+class OpensearchDashboardsDependencyModel(BaseModel):
+    """Model for OpensearchDashboards Operator dependencies."""
+
+    osd_upstream: DependencyModel
+
+
+class ODUpgradeEvents(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: "OpensearchDashboardsCharm", **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    def post_upgrade_check(self) -> None:
```

Comment:
I recommend we add a `workload_version` file in this charm. At `post_upgrade_check`, you can check the version of your snap currently installed vs. your workload. If this is not, e.g. a major upgrade, then we can proceed.

Reply:
To summarize a live discussion, we found that this "upgrades v2" feature is rather to be used when moving to "v2"